### PR TITLE
Builtin Map domain

### PR DESCRIPTION
**Pull request by ttuegel**
Defines the builtin `MAP.Map` domain and associated operations. An efficient internal representation is used.

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`
- [ ] Style conformance: `stylish-haskell`

---

### CODE REVIEW

File: src/main/haskell/kore/src/Kore/Builtin/Builtin.hs [hunk 1]
```diff
@@ -548,22 +537,49 @@ functionEvaluator impl =
...
+{- | Look up the symbol hooked to the named builtin in the provided module.
+ -}
+lookupSymbol
+    :: String
+    -- ^ builtin name
+    -> KoreIndexedModule attrs
+    -> Either (Error e) (SymbolOrAlias Object)
+lookupSymbol builtinName indexedModule
+  = do
+    symbolOrAliasConstructor <- IndexedModule.resolveHook indexedModule builtinName
```
Comment (virgil-serbanuta) [inline src/main/haskell/kore/src/Kore/Builtin/Builtin.hs]: Line length

File: src/main/haskell/kore/src/Kore/Builtin/Map.hs [hunk 2]
```diff
@@ -71,89 +89,245 @@ sortDeclVerifiers = HashMap.fromList [ (sort, Builtin.verifySortDecl) ]
...
-    evalBind0 _ _ _ = notApplicableFunctionEvaluator
+    anySort :: Builtin.SortVerifier
+    anySort = const $ const $ Right ()
+
+type Builtin = Map (CommonPurePattern Object) (CommonPurePattern Object)
+
+{- | Abort function evaluation if the argument is not a Map domain value.
+
+    If the operand pattern is not a domain value, the function is simply
+    'NotApplicatible'. If the operand is a domain value, but not represented
```
Comment (virgil-serbanuta) [inline src/main/haskell/kore/src/Kore/Builtin/Map.hs]: NotApplicable

File: src/main/haskell/kore/src/Kore/Builtin/Map.hs [hunk 3]
```diff
@@ -71,89 +89,245 @@ sortDeclVerifiers = HashMap.fromList [ (sort, Builtin.verifySortDecl) ]
...
+expectBuiltinDomainMap
+    :: Monad m
+    => String  -- ^ Context for error message
+    -> CommonPurePattern Object  -- ^ Operand pattern
+    -> ExceptT (AttemptedFunction Object Kore.Variable) m Builtin
+expectBuiltinDomainMap ctx =
+    \case
+        Kore.DV_ _ domain ->
+            case domain of
+                Kore.BuiltinDomainMap _map -> return _map
```
Comment (virgil-serbanuta) [inline src/main/haskell/kore/src/Kore/Builtin/Map.hs]: Same question as above: Why not map_ or map' or something else? The same for all other occurrences of _map, _key, _value and maybe others.

File: src/main/haskell/kore/src/Kore/IndexedModule/IndexedModule.hs [hunk 4]
```diff
@@ -498,134 +513,151 @@ indexModuleObjectSentence
...
+                implicitModule
+                importingModules
+                nameToModule
+                (indexedModules, indexedModule)
+                (SentenceSymbolSentence
+                    SentenceSymbol
+                        { sentenceSymbolSymbol = Symbol
+                            { symbolConstructor = Id
+                                { getId = metaNameForObjectSort (getId sentenceSortName)
+                                , idLocation = AstLocationLifted (idLocation sentenceSortName)
```
Comment (virgil-serbanuta) [inline src/main/haskell/kore/src/Kore/IndexedModule/IndexedModule.hs]: Line length, here and above.

File: src/main/haskell/kore/src/Logic/Matching/Prover/Command.hs [hunk 5]
```diff
@@ -74,12 +76,24 @@ parseCommand pIx pFormula pDerivation
...
-instance (Pretty ix, Pretty formula, Pretty (rule ix)) => Pretty (Command ix rule formula) where
-  pretty (Add ix formula)              =  pretty(  "add"::Text)<+>pretty ix<+>colon<+>pretty formula
-  pretty (Prove ix rule)               =  pretty("prove"::Text)<+>pretty ix<+>colon
-                                      <+> pretty(   "by"::Text)<+>pretty rule
-  pretty (AddAndProve ix formula rule) =  pretty(  "add"::Text)<+>pretty ix<+>colon<+>pretty formula
-                                      <+> pretty(   "by"::Text)<+>pretty rule
-  pretty Undo                          =  pretty( "undo"::Text)
-  pretty Show                          =  pretty( "show"::Text)
-  pretty Help                          =  pretty( "help"::Text)
+instance (Pretty ix, Unparse formula, Pretty (rule ix)) => Pretty (Command ix rule formula) where
```
Comment (virgil-serbanuta) [inline src/main/haskell/kore/src/Logic/Matching/Prover/Command.hs]: Line length.

File: src/main/haskell/kore/src/Logic/Matching/Rules/Minimal/Syntax.hs [hunk 6]
```diff
@@ -108,22 +109,22 @@ parseMLRule pLabel pVar pTerm pIx =
...
        => Pretty (MLRule label var term hyp) where
   pretty proofRule = let
       rule :: Text -> [Doc ann] -> Doc ann
       rule prefix args = pretty prefix <> tupled args
     in case proofRule of
-    Propositional1 p1 p2 -> rule "propositional1" [pretty p1,pretty p2]
-    Propositional2 p1 p2 p3 -> rule "propositional2" [pretty p1,pretty p2,pretty p3]
-    Propositional3 p1 p2 -> rule "propositional3" [pretty p1, pretty p2]
+    Propositional1 p1 p2 -> rule "propositional1" [unparse p1,unparse p2]
+    Propositional2 p1 p2 p3 -> rule "propositional2" [unparse p1,unparse p2,unparse p3]
```
Comment (virgil-serbanuta) [inline src/main/haskell/kore/src/Logic/Matching/Rules/Minimal/Syntax.hs]: Line length here and below.

File: src/main/haskell/kore/test/Test/Kore/Parser/Parser.hs [hunk 7]
```diff
@@ -463,7 +463,7 @@ domainValuePatternParserTests =
     parseTree korePatternParser
         [ success "\\dv{s1}(\"a\")"
             ( patternPureToKore
-            $ DV_ (sortVariableSort "s1") (StringLiteral_ (StringLiteral "a"))
+            $ DV_ (sortVariableSort "s1") (BuiltinDomainPattern (StringLiteral_ "a"))
```
Comment (virgil-serbanuta) [inline src/main/haskell/kore/test/Test/Kore/Parser/Parser.hs]: Line length.

**Comment by traiansf** [comment]
@ttuegel nevermind my prior comment about addressing suggestions. Seems you have in the meanwhile :-)

**Pull request merged**